File: Stacks & Queues/#503 - Next greater element 2 - Medium/nextGreaterElements.java
class Solution {
    public int[] nextGreaterElements(int[] nums) {
        int n = nums.length;
        Stack<Integer> st = new Stack<>();
        int ans[] = new int[n];
        // Initialize stack for circular behavior
        for (int i = n - 1; i >= 0; i--) {
            st.push(nums[i]);
        }
        for (int i = n - 1; i >= 0; i--) {
            while (!st.isEmpty() && nums[i] >= st.peek()) {
                st.pop();
            }
            if (st.isEmpty())
                ans[i] = -1;
            else
                ans[i] = st.peek();
            st.push(nums[i]);
        }
        return ans;
    }
}

File: Stacks & Queues/#503 - Next greater element 2 - Medium/nextGreaterElements.py
class Solution:
    def nextGreaterElements(self, nums: List[int]) -> List[int]:
        n = len(nums)
        stack = []
        ans = [0] * n
        # Preload stack with reversed nums for circular behavior
        for i in range(n - 1, -1, -1):
            stack.append(nums[i])
        for i in range(n - 1, -1, -1):
            while stack and nums[i] >= stack[-1]:
                stack.pop()
            if not stack:
                ans[i] = -1
            else:
                ans[i] = stack[-1]
            stack.append(nums[i])
        return ans
